mindmeld/active_learning/heuristics.py

```diff
@@ -0,0 +1,621 @@
+                corresponding indices in the probability vector. Used for intent-level KLD.
+        Returns:
+            divergences: List of divergence values for each query compared to the average
+                within-intent probability distribution.
+        """
+        divergences = []
+        # Calculate q_d
+        q_d = {d: [] for d in domain_indices}
+        for row in avg_preds:
+            domain = KLDivergenceSampling.get_domain(domain_indices, row)
+            q_d[domain].append(row)
+        for pred in preds_multi:
+            # Calculate p_d
+            p_d = {d: [] for d in domain_indices}
+            for row in pred:
+                domain = KLDivergenceSampling.get_domain(domain_indices, row)
+                p_d[domain].append(row)
+            # Calculate Divergence Scores by Domain
+            divergence_scores_by_domain = {d: [] for d in domain_indices}
+            for domain in domain_indices:
+                if len(p_d[domain]) > 0 and len(q_d[domain]) > 0:
+                    divergence_scores_by_domain[domain] = scipy_entropy(
+                        np.array(p_d[domain]),
+                        np.array(q_d[domain]),
+                        axis=1,
+                        base=ENTROPY_LOG_BASE,
+                    )
+                else:
+                    divergence_scores_by_domain[domain] = 0
+            # Reorder Divergence Scores by Domain to Original order
+            single_pred_divergence_scores = []
+            domain_counter = {d: 0 for d in domain_indices}
+            for row in pred:
+                domain = KLDivergenceSampling.get_domain(domain_indices, row)
+                single_pred_divergence_scores.append(
+                    divergence_scores_by_domain[domain][domain_counter[domain]]
+                )
+                domain_counter[domain] += 1
+            divergences.append(single_pred_divergence_scores)
+        return divergences
+
+    @staticmethod
+    def get_domain(domain_indices, row):
+        """Get the domain for a given probability row, inferred based on the non-zero values.
+        Args:
+            preds_multi (List[List[float]]): Probability scores for each data point for each
+                class from multiple classifiers.
+            avg_preds (List[float]): Average probability vector across all intents.
+            domain_indices (Dict[str, tuple(int, int)]): A mapping between domains (str) to the
+                corresponding indices in the probability vector. Used for intent-level KLD.
+            row (List[List[float]]): A single row representing a queries probability distrubition.
+        Returns:
+            domain (str): The domain that the given row belongs to.
+        Raises:
+            AssertionError: If a row does not have an associated domain.
+        """
+        if np.sum(row) == 0:
+            row = [1 / len(row)] * len(row)
```

Comment:
shouldnt this be 1?

Reply:
This aimed to create a uniform distribution of 1/len(row), for example: [1/3, 1/3, 1/3]. Changed to the following:

`row = list(np.repeat(1 / len(row), repeats=len(row)))`